File: src/application/useCases/addTransaction.ts
import { assertHalalas, type Currency, type Halalas } from '../../domain/money'
import { isValidIsoDate } from '../../domain/period'
import { isConsistentWithObservedDirection, liquidityOf, type EconomicKind } from '../../domain/entities/economicKind'
import type { Id, Transaction } from '../../domain/entities/types'
import type { Clock, IdGenerator, TransactionRepository, WalletRepository } from '../ports/repositories'

/**
 * AddTransaction — إضافة عملية يدويًا.
 *
 * `spec/01`: «الإضافة: **فاتورة/شراء**، حركة أموال، كشف حساب، شخص جديد».
 *
 * ⚠️ الحاجة اكتشفها المالك بالاستعمال: اشترى قهوة بالكاش ولم يجد أين
 * يسجّلها — زر الإضافة كان يفتح الاستيراد وحده. والكاش تحديدًا **لا يصل
 * من أي كشف**، فبدون هذا المسار يبقى صرف الكاش غير مرصود بالكامل.
 *
 * الفرق عن الاستيراد: هنا المستخدم **يعرف** ما فعله، فالنوع الاقتصادي
 * يُحدَّد ويُؤكَّد فورًا — لا يمر بدورة «غير محدد ثم اقتراح ثم تأكيد».
 */

export class AddTransactionError extends Error {}

export interface NewTransactionInput {
  /** موجب دائمًا. الاتجاه من النوع الاقتصادي. */
  amountMinor: Halalas
  currency?: Currency
  occurredAt: string
  walletId: Id
  economicKind: EconomicKind
  merchantName: string
  categoryId?: Id
  note?: string
  isCashTagged?: boolean
  excludedFromBudget?: boolean
}

export interface AddTransactionDeps {
  txns: TransactionRepository
  wallets: WalletRepository
  ids: IdGenerator
  clock: Clock
}

export function makeAddTransaction(deps: AddTransactionDeps) {
  return async function add(input: NewTransactionInput): Promise<Transaction> {
    if (!Number.isInteger(input.amountMinor)) {
      throw new AddTransactionError('المبلغ لازم يكون بالهللة كعدد صحيح')
    }
    if (input.amountMinor <= 0) {
      throw new AddTransactionError('المبلغ لازم يكون أكبر من صفر. الاتجاه بيتحدد من نوع العملية.')
    }
    assertHalalas(input.amountMinor, 'مبلغ العملية')

    if (!isValidIsoDate(input.occurredAt)) {
      throw new AddTransactionError('التاريخ مش صالح')
    }

    const wallet = await deps.wallets.findById(input.walletId)
    if (!wallet) throw new AddTransactionError('اختار محفظة موجودة')

    if (input.economicKind === 'unclassified') {
      throw new AddTransactionError('لازم تحدد نوع العملية — انت عارف اشتريت ولا حوّلت')
    }

    /*
     * الاتجاه الملاحظ يُشتق من النوع هنا — عكس الاستيراد تمامًا.
     * في الاستيراد الكشف يعطي الاتجاه والنوع مجهول؛ هنا المستخدم يعطي
     * النوع والاتجاه نتيجته. `internal_transfer` يُعامل صادرًا من محفظة
     * المصدر، وطرفه الثاني عمل لم يُبنَ بعد (ARCHITECTURE §14).
     */
    const liquidity = liquidityOf(input.economicKind)
    const observedDirection: 'in' | 'out' = liquidity === 'in' ? 'in' : 'out'

    if (!isConsistentWithObservedDirection(input.economicKind, observedDirection)) {
      throw new AddTransactionError('النوع ده ما يتوافقش مع اتجاه الحركة')
    }

    const name = input.merchantName.trim()
    if (name.length > 120) {
      throw new AddTransactionError('اسم المتجر أطول من 120 حرف')
    }
    if ((input.note ?? '').length > 1000) {
      throw new AddTransactionError('الملاحظة أطول من 1000 حرف')
    }

    const now = deps.clock.nowIso()
    const transaction: Transaction = {
      id: deps.ids.next('txn'),
      occurredAt: input.occurredAt,
      datePrecision: 'day',
      // العمليات اليدوية بعد أي عملية مستوردة في نفس اليوم
      sourceOrder: 9_000_000,
      economicKind: input.economicKind,
      // ✅ مؤكَّد فورًا: المستخدم اختاره بنفسه، فلا يُكتب فوقه آليًا (spec/05)
      economicKindConfirmed: true,
      observedDirection,
      amountMinor: input.amountMinor,
      currency: input.currency ?? wallet.currency,
      walletId: wallet.id,
      categoryConfirmed: input.categoryId !== undefined,
      excludedFromBudget: input.excludedFromBudget ?? false,
      reviewState: input.categoryId ? 'confirmed' : 'needs_review',
      // الكاش يُوسم تلقائيًا حين تكون المحفظة كاش — الوسم شارة عرض لا مبلغ
      isCashTagged: input.isCashTagged ?? wallet.kind === 'cash',
      rawMerchantName: name || 'بلا اسم',
      createdAt: now,
      updatedAt: now,
    }

    if (input.categoryId) transaction.categoryId = input.categoryId
    if (input.note?.trim()) transaction.note = input.note.trim()

    await deps.txns.saveMany([transaction])
    return transaction
  }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/usecase/AddTransaction.kt
package app.masroufy.usecase

import app.masroufy.core.Currency
import app.masroufy.core.Direction
import app.masroufy.core.EconomicKind
import app.masroufy.core.Halalas
import app.masroufy.core.Id
import app.masroufy.core.Liquidity
import app.masroufy.core.ReviewState
import app.masroufy.core.Transaction
import app.masroufy.core.assertHalalas
import app.masroufy.core.isConsistentWithObservedDirection
import app.masroufy.core.isValidIsoDate
import app.masroufy.core.jsTrim
import app.masroufy.core.ruleFor
import app.masroufy.port.Clock
import app.masroufy.port.IdGenerator
import app.masroufy.port.TransactionRepository
import app.masroufy.port.WalletRepository

/**
 * AddTransaction — نقل `addTransaction.ts`: إضافة عملية يدويًا (القهوة بالكاش ما بتوصلش من أي كشف).
 * الفرق عن الاستيراد: هنا المستخدم **عارف** عمل إيه، فالنوع بيتحدد وبيتأكد فورًا،
 * والاتجاه الملاحظ **بيشتق من النوع** — عكس الاستيراد بالظبط.
 * («المبلغ عدد صحيح» مش متفحصة هنا — نظام أنواع كوتلن بيمنعها وقت الترجمة.)
 */

data class NewTransactionInput(
    /** موجب دايمًا. الاتجاه من النوع الاقتصادي. */
    val amountMinor: Halalas,
    val currency: Currency? = null,
    val occurredAt: String,
    val walletId: Id,
    /** المحفظة المستقبِلة — إلزامية للتحويل الداخلي وحده. */
    val transferToWalletId: Id? = null,
    val economicKind: EconomicKind,
    val merchantName: String,
    val categoryId: Id? = null,
    val note: String? = null,
    val isCashTagged: Boolean? = null,
    val excludedFromBudget: Boolean? = null,
)

data class AddTransactionDeps(
    val txns: TransactionRepository,
    val wallets: WalletRepository,
    val ids: IdGenerator,
    val clock: Clock,
)

class AddTransaction(private val deps: AddTransactionDeps) {
    suspend fun add(input: NewTransactionInput): Transaction {
        if (input.amountMinor <= 0) {
            throw IllegalArgumentException("المبلغ لازم يكون أكبر من صفر. الاتجاه بيتحدد من نوع العملية.")
        }
        assertHalalas(input.amountMinor, "مبلغ العملية")

        if (!isValidIsoDate(input.occurredAt)) throw IllegalArgumentException("التاريخ مش صالح")

        val wallet = deps.wallets.findById(input.walletId) ?: throw IllegalArgumentException("اختار محفظة موجودة")

        if (input.economicKind == EconomicKind.UNCLASSIFIED) {
            throw IllegalArgumentException("لازم تحدد نوع العملية — انت عارف اشتريت ولا حوّلت")
        }

        // الاتجاه الملاحظ بيشتق من النوع هنا — في الاستيراد الكشف بيدي الاتجاه والنوع مجهول
        val observedDirection = if (ruleFor(input.economicKind).liquidity == Liquidity.IN) Direction.IN else Direction.OUT

        if (!isConsistentWithObservedDirection(input.economicKind, observedDirection)) {
            throw IllegalArgumentException("النوع ده ما يتوافقش مع اتجاه الحركة")
        }

        /*
         * التحويل الداخلي **لازم له طرفين** (spec/02) — من غير المحفظة المستقبِلة
         * الفلوس بتختفي من الحساب بلا أثر. بيترفض بدل ما يتحفظ ناقص.
         */
        var transferTo: Id? = null
        if (input.economicKind == EconomicKind.INTERNAL_TRANSFER) {
            val targetId = input.transferToWalletId
                ?: throw IllegalArgumentException("التحويل الداخلي لازم تحدد راح لأنهي محفظة")
            if (targetId == input.walletId) throw IllegalArgumentException("مينفعش تحوّل من محفظة لنفسها")
            val target = deps.wallets.findById(targetId) ?: throw IllegalArgumentException("المحفظة المستقبِلة مش موجودة")
            if (target.currency != wallet.currency) {
                throw IllegalArgumentException("التحويل بين عملتين مختلفتين محتاج سعر صرف موثّق — لسه مش مدعوم (spec/02)")
            }
            transferTo = target.id
        } else if (input.transferToWalletId != null) {
            throw IllegalArgumentException("المحفظة المستقبِلة تتحدد للتحويل الداخلي بس")
        }

        val name = jsTrim(input.merchantName)
        if (name.length > 120) throw IllegalArgumentException("اسم المتجر أطول من 120 حرف")
        if ((input.note ?: "").length > 1000) throw IllegalArgumentException("الملاحظة أطول من 1000 حرف")

        val now = deps.clock.nowIso()
        val note = input.note?.let { jsTrim(it) }?.takeIf { it.isNotEmpty() }
        val transaction = Transaction(
            id = deps.ids.next("txn"),
            occurredAt = input.occurredAt,
            datePrecision = "day",
            // العمليات اليدوية بعد أي عملية مستوردة في نفس اليوم
            sourceOrder = 9_000_000,
            economicKind = input.economicKind,
            // ✅ مؤكد فورًا: المستخدم اختاره بنفسه، فما يتكتبش فوقه آليًا (spec/05)
            economicKindConfirmed = true,
            observedDirection = observedDirection,
            amountMinor = input.amountMinor,
            currency = input.currency ?: wallet.currency,
            walletId = wallet.id,
            categoryConfirmed = input.categoryId != null,
            excludedFromBudget = input.excludedFromBudget ?: false,
            reviewState = if (input.categoryId != null) ReviewState.CONFIRMED else ReviewState.NEEDS_REVIEW,
            // الكاش بيتوسم تلقائيًا لما المحفظة كاش — الوسم شارة عرض مش مبلغ
            isCashTagged = input.isCashTagged ?: (wallet.kind == "cash"),
            rawMerchantName = name.ifEmpty { "بلا اسم" },
            createdAt = now,
            updatedAt = now,
            transferToWalletId = transferTo,
            categoryId = input.categoryId,
            note = note,
        )

        deps.txns.saveMany(listOf(transaction))
        return transaction
    }
}
